### PR TITLE
feat: decoder kogpt2로 변경

### PR DESCRIPTION
**Pull request by gkwjdtk09**
decoder를 kogpt2로 변경 시도하였으나 코드 혼선이 있어서 기존 코드에서 다시 작업하려고 합니다.
huggingface 등에 올라온 pretrained 된 모델들을 어떻게 적용하고 뭘 수정해줘야하는지 아직 어렵습니다.
내일은 꼭 적용할 수 있도록 하겠습니다.

### CODE REVIEW

**Comment by gkwjdtk09** [comment]
## 변경사항
- KoGPT2와 ResNet50을 활용한 이미지 캡셔닝 파이프라인을 구현했습니다.
- EncoderCNN (ResNet 기반)과 Decoder (KoGPT2 기반) 모델 구현
- 작은 데이터셋 기반 테스트 완료했습니다. 성능 개선 예정입니다. 

## 진행 계획
- 현재 결과는 반복되는 토큰 또는 불완전한 문장이 포함되는데, 샘플링 방식 때문이라고 예측하고 있습니다.
image
- 성능 개선을 위해 Inference 코드를 개선할 예정입니다. (Top-k / Top-p 샘플링 방식이 있다고 해서 알아보고 적용해볼 계획입니다.)
- BLEU, ROUGE 등의 평가 지표 적용할 예정입니다.

File: JeongEunPark/modeling/dataset.py [hunk 1]
```diff
@@ -3,42 +3,59 @@
...
         except Exception:
-            img = Image.new("RGB", (224, 224))
-        img = self.transform(img)
-        caption_tensor = self.get_input_ids(caption)
\ No newline at end of file
+            image = Image.new("RGB", (224, 224))
+        image = self.transform(image)
+
+        caption = self.captions_list[idx]
+        caption_with_token = f"</s>{caption}</s>"
```
Comment (saitros) [inline JeongEunPark/modeling/dataset.py]: </s> 이렇게 하드코딩으로 하기보다는 tokenizer 에서 직접 가져오는 방식으로 하는게 dataset 으로 사용하면서 다른 모델이랑 호환성을 높일 수 있습니다.

File: JeongEunPark/modeling/inference.py [hunk 2]
```diff
@@ -1,63 +1,55 @@
...
+
+            if next_token == tokenizer.eos_token_id:
+                break
+            generated.append(next_token)
+
+    # 토큰 시퀀스를 텍스트로 변환
+    caption = tokenizer.decode(generated, skip_special_tokens=True)
+    return caption
+
+    # def build_vocab_from_csv(csv_path, max_samples=100):
```
Comment (saitros) [inline JeongEunPark/modeling/inference.py]: 주석 안쓰는건 삭제해주세요

File: JeongEunPark/modeling/model.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        )
+
+        return outputs
+        # embeddings = self.embed(captions[:, :-1])  # [B, T-1, embed_size]
+        # embeddings = torch.cat((features.unsqueeze(1), embeddings), 1)
+        # lstm_out, _ = self.lstm(embeddings)
+        # outputs = self.linear(self.dropout(lstm_out))
+        # return outputs
+
+    # def decode_step(self, input_token, hidden):
```
Comment (saitros) [inline JeongEunPark/modeling/model.py]: 안쓰는건 최대한 삭제해주세요

File: JeongEunPark/modeling/train.py [hunk 4]
```diff
@@ -4,57 +4,90 @@
...
 
-def train_model(encoder, decoder, train_loader, vocab_size, device,
-                num_epochs=10, learning_rate=1e-3):
+    # for images, captions in tqdm(dataloader, desc=">>> Training", leave=False):
+    #     images, captions = images.to(device), captions.to(device)
 
-    criterion = nn.CrossEntropyLoss(ignore_index=0)  # 0은 padding
-    encoder_optimizer = optim.Adam(encoder.parameters(), lr=learning_rate)
-    decoder_optimizer = optim.Adam(decoder.parameters(), lr=learning_rate)
+    #     encoder_optimizer.zero_grad()
```
Comment (saitros) [inline JeongEunPark/modeling/train.py]: 안쓰는 주석은 삭제

File: JeongEunPark/modeling/dataset.py (hunk 1 above)
Comment (gkwjdtk09) [reply JeongEunPark/modeling/dataset.py]: 넵 알겠습니다.

**Comment by gkwjdtk09** [comment]
- train_model() 함수에 Early Stopping 적용 (patience=3)
- AdamW 옵티마이저 적용 + learning rate (5e-5)설정
- ResNet encoder는 torch.no_grad()로 Freeze, decoder만 학습되도록 처리
- optimizer 설정 시 encoder.parameters()는 제외함

- 추론 방식 Beam Search 방식 추가 (beam_width=5)

- 불필요한 주석 삭제

해당 내용 수정해서 백그라운드로 학습 해놓았습니다.

**Comment by gkwjdtk09** [comment]
### 변경 사항
- `EarlyStopping` 클래스 구현 및 적용 → trainloss가 0.1 이하로 내려갔을 시 val loss 기반 조기 종료
- `ReduceLROnPlateau` 스케줄러 적용 → mode='min', factor=0.5, patience=10, verbose=True
-> loss 그래프 확인 후 CosineAnnealingLR 또한 적용해볼 예정
- 학습 후 train/validation loss 그래프(`loss_plot.png`) 저장 기능 추가
- tokenizer의 BOS/EOS 토큰을 하드코딩 대신 tokenizer에서 직접 불러오도록 수정

---

### 파일 설명 
- `early_stopping.py`: 조기 종료 기능 클래스화 및 적용  
- `train.py`: val_dataloader 추가, loss 기록 및 시각화 로직 포함  
- `main.py`: tokenizer special token 명시적 설정 및 split reproducibility 고정  

현재 변경된 코드로 train (70epoch) 진행중입니다.

**Comment by gkwjdtk09** [comment]
아래의 이미지는 변경된 코드로 train 70epoch 진행한 loss plot 입니다.
validation loss가 점점 증가하는 것으로 보아 과적합이 의심됩니다.
patience를 높여서 더 학습이 천천히 되게 하려고 합니다.
image

**Comment by gkwjdtk09** [comment]
scheduler 의  patience를 15로 설정하고, 80에폭 학습한 결과 아래와 같은 loss가 나왔습니다.
loss_plot_ep80_p15
train loss는 계속 줄기 때문에 early stopping이 되지 않은 것으로 보입니다.
아래는 모델을 가지고 추론한 결과입니다.
image
